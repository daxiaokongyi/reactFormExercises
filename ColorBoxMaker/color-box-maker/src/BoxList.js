import React, {useState} from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import uuid from 'react-uuid';

const BoxList = () => {
    // set usestate for box
    const [boxes, setBoxes] = useState([]);

    // add remove function on the box
    const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    }

    const addBox = box => {
        setBoxes(boxes => [...boxes, box]);
    }

    const renderBoxes = boxes.map(({backgroundColor, width, height, id}) => (
        <Box backgroundColor = {backgroundColor} 
                key = {id}
                id = {id}
                width = {width} 
                height = {height} 
                handleRemove = {removeBox}
        />)
    );

    return (
        <div>
            {renderBoxes}
            <NewBoxForm addBox = {addBox}/>
        </div>
    )
}

export default BoxList;