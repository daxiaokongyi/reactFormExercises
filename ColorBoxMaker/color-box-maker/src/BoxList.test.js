import React from 'react';
import BoxList from './BoxList';
import {render, fireEvent, wait} from '@testing-library/react';

const addBox = (boxList, height = "100" , width = "100", color = "green") => {
    const backgroundColorInput = boxList.getByLabelText("Background Color:");
    const heightInput = boxList.getByLabelText("Height:");
    const widthInput = boxList.getByLabelText("Width:");
    
    fireEvent.change(backgroundColorInput, {target: {value: color}});
    fireEvent.change(heightInput, {target: {value: height}});
    fireEvent.change(widthInput, {target: {value: width}});

    const button = boxList.getByText("Add a new box");
    fireEvent.click(button);
} 

// smoke test
it("renders without crashing", () => {
    render(<BoxList/>);
});

// snapshot test
it("matches snapshot", () => {
    const {asFragment} = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
})

// test if adding a box works
it("should be able to add a box", () => {
    const boxList = render(<BoxList/>);

    // no box yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();

    // add a box by using addBox function
    addBox(boxList);

    // expect to see a box
    const removeBtn = boxList.getByText("X");
    expect(removeBtn).toBeInTheDocument();
    expect(removeBtn.previousSibling).toHaveStyle(`
        width: 100px;
        height: 100px;
        backgroundColor: green 
    `);
    // expect form becomes empty
    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);
})

it("should be able to remove a box", () => {
    const boxList = render(<BoxList/>);
    // add a box into the boxlist
    addBox(boxList);
    const removeBtn = boxList.getByText("X");
    // check if remove button exists
    expect(removeBtn).toBeInTheDocument();
    // click the remove button to remove the added box
    fireEvent.click(removeBtn);
    expect(removeBtn).not.toBeInTheDocument();
})