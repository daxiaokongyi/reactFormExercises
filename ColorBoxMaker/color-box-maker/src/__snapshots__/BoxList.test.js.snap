// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="backgroundColor"
        >
          Background Color: 
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          type="text"
          value=""
        />
        <label
          for="height"
        >
          Height: 
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
        <label
          for="width"
        >
          Width: 
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
        <button>
          Add a new box
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
